### PR TITLE
feat: add information_schema.key_column_usage

### PR DESCRIPTION
**Pull request by dimbtp**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- add `information_schema.key_column_usage` table

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

## Refer to a related PR or issue link (optional)

#2931

### CODE REVIEW

File: src/catalog/src/information_schema/key_column_usage.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+use store_api::storage::TableId;
+
+use super::KEY_COLUMN_USAGE;
+use crate::error::{
+    CreateRecordBatchSnafu, InternalSnafu, Result, UpgradeWeakCatalogManagerRefSnafu,
+};
+use crate::information_schema::InformationTable;
+use crate::CatalogManager;
+
+pub(super) struct InformationSchemaKeyColumnUsage {
```
Comment (killme2008) [inline src/catalog/src/information_schema/key_column_usage.rs:39]:
```suggestion
/// The virtual table implementation for `information_schema. KEY_COLUMN_USAGE`.
pub(super) struct InformationSchemaKeyColumnUsage {
```

Comment (killme2008) [reply src/catalog/src/information_schema/key_column_usage.rs:39]: We'd like to always add a description for structures.

File: src/catalog/src/information_schema/key_column_usage.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+                    let keys = &table.table_info().meta.primary_key_indices;
+                    let schema = table.schema();
+
+                    for (idx, column) in schema.column_schemas().iter().enumerate() {
+                        let constraint_name = if column.is_time_index() {
+                            "TIME INDEX"
+                        } else if keys.contains(&idx) {
+                            "PRIMARY"
+                        } else {
+                            // TODO: foreign constraint
```
Comment (killme2008) [inline src/catalog/src/information_schema/key_column_usage.rs:226]:
```suggestion
                            // TODO(dimbtp): foreign constraint not supported yet
```

File: src/catalog/src/information_schema/key_column_usage.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        constraint_name: &str,
+        table_catalog: &str,
+        table_schema: &str,
+        table_name: &str,
+        column_name: &str,
+        ordinal_position: u32,
+        position_in_unique_constraint: Option<u32>,
+        referenced_table_schema: Option<&str>,
+        referenced_table_name: Option<&str>,
+        referenced_column_name: Option<&str>,
```
Comment (killme2008) [inline src/catalog/src/information_schema/key_column_usage.rs:268]: These values are always `None`, so we don't need to pass them, and just insert them in `add_key_column_usage` directly. After this refactoring, we can remove `#[allow(clippy::too_many_arguments)]`.

File: src/catalog/src/information_schema/key_column_usage.rs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+                }
+            }
+        }
+
+        self.finish()
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn add_key_column_usage(
+        &mut self,
```
Comment (killme2008) [inline src/catalog/src/information_schema/key_column_usage.rs:256]: The `constraint_catalog` and `table_catalog` are always `def` by default, so we don't need to pass them, and just insert the `def` value directly in this function.

File: src/catalog/src/information_schema/key_column_usage.rs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+                        self.add_key_column_usage(
+                            "def",
+                            &schema_name,
+                            constraint_name,
+                            "def",
+                            &schema_name,
+                            &table_name,
+                            &column.name,
+                            idx as u32 + 1,
```
Comment (killme2008) [inline src/catalog/src/information_schema/key_column_usage.rs:238]: The `ordinal_position` is the column's position **within the constraint, not the column's position within the table**.

 So I think we can add a mutable value to index the primary columns just like:

```rust
let keys = &table.table_info().meta.primary_key_indices;
                    let schema = table.schema();

let primary_key_index = 0;

for (idx, column) in schema.column_schemas().iter().enumerate() {
    let constraint_name = if column.is_time_index() {
          "TIME INDEX"
    } else if keys.contains(&idx) {
          primary_key_index = primary_key_index + 1;
          "PRIMARY"
    } else {
           // TODO: foreign constraint
           continue;
    };

    self.add_key_column_usage(
                            "def",
                            &schema_name,
                            constraint_name,
                            "def",
                            &schema_name,
                            &table_name,
                            &column.name,
                            primary_key_index,
                            None,
                            None,
                            None,
                            None,
                        );
}



```

**Review by killme2008 — COMMENTED** [review]
Great job! Thank you.

A few suggestions are attached, please take a look. And i think it's better to add some sql to test this table to `tests/cases/standalone/common/system/information_schema.sql`, just like

https://github.com/GreptimeTeam/greptimedb/blob/6070e880776e0216796a0155420ed7dafb5d18a0/tests/cases/standalone/common/system/information_schema.sql#L69

We can ensure the virutal table results are correct.

**Comment by killme2008** [comment]
BTW. After #3054 merged, some conflicts have to be resolved right now.

File: src/catalog/src/information_schema/key_column_usage.rs (hunk 3 above)
Comment (dimbtp) [reply src/catalog/src/information_schema/key_column_usage.rs:268]: These values are not `None` when there is a `foreign key constraint`. So should I put a `TODO: not None for foreign key` and set these values as `None` now?

File: src/catalog/src/information_schema/key_column_usage.rs (hunk 5 above)
Comment (dimbtp) [reply src/catalog/src/information_schema/key_column_usage.rs:238]: Got it.

Should I count one column multiple times when there are multiple key constraints? I find that there is only `time index` and `primary` now, and `time index` is treated as `primary`, do I need to leave a `TODO` here?

File: src/catalog/src/information_schema/key_column_usage.rs (hunk 3 above)
Comment (killme2008) [reply src/catalog/src/information_schema/key_column_usage.rs]: Yes, I think we can add them back when we support foreign key constraint in the future.

**Review by killme2008 — APPROVED** [review]
LGTM. Thank you very much!

**Review by sunng87 — APPROVED** [review]

**Pull request merged**